tee-apps/sealed-bid-auction/src/core/AuctionService.ts

```diff
@@ -0,0 +1,86 @@
+import type {PriceListItem} from "./types.ts";
+
+type Price = {
+    amountOut: bigint;
+    solverAddress: string;
+}
+
+export class AuctionService {
+    constructor(private readonly solverPricebook: SolverPriceBook) {}
+
+    public auction(request: AuctionRequest): AuctionQuote | null {
+        const pricesForAskedTokens = this.findPricesForAskedTokens(request);
+
+        const bestPrice = this.chooseBestPrice(pricesForAskedTokens);
+
+        if (!bestPrice) {
+            throw new Error("No quote found for this pair");
+        } else {
+            return {
+                id: request.id,
+                request: request as Omit<AuctionRequest, "id">,
+                amountOut: bestPrice.amountOut,
+                solverAddress: bestPrice.solverAddress,
+                timestamp: Date.now()
+            };
+        }
+    }
+
+    private getFinalIntervalIndex(priceListItem: PriceListItem, amount: bigint): number | undefined {
+        return priceListItem.intervals.findIndex(interval => interval.range.min <= amount && interval.range.max >= amount);
+    }
+
+    private chooseBestPrice(pricesForAskedTokens: Immutable.List<Price>): Price | null {
+        let bestPrice = null;
+
+        if (!pricesForAskedTokens.isEmpty()) {
+            let bestPrice: Price = {
+                amountOut: -1n,
+                solverAddress: "0xdeadbeef"
+            };
+
+            pricesForAskedTokens.forEach(price => {
+                if (price.amountOut > bestPrice.amountOut) bestPrice = price;
+            });
+        }
+
+        return bestPrice;
+    }
+
+    private findPricesForAskedTokens(request: AuctionRequest): Immutable.List<Price> {
+        return this.solverPricebook.getCurrentPrices().flatMap(
+            priceItems => priceItems.filter(
+                priceItem =>
+                    priceItem.srcTokenAddresses.includes(request.srcTokenAddress) &&
+                    priceItem.dstTokenAddress.includes(request.dstTokenAddress) &&
+                    this.getFinalIntervalIndex(priceItem, request.amountIn) !== undefined
+            ).map(priceItem => {
+                return {
+                    amountOut: this.calculateAmountOut(priceItem, request.amountIn),
+                    solverAddress: priceItem.solverAddress
+                };
+            })
+        )
+    }
+
+    private calculateAmountOut(priceItem: PriceListItem, amountIn: bigint): bigint {
+        let currentInterval = 0;
+        let amountOut = 0n;
+        const finalIndex = this.getFinalIntervalIndex(priceItem, amountIn)!;
+
+        do {
+            const currInterval = priceItem.intervals[currentInterval]!;
+            if (currInterval !== priceItem.intervals[finalIndex]!) {
+                amountOut += currInterval.price * (currInterval.range.max - currInterval.range.min);
+            } else {
+                amountOut += currInterval.price * (amountIn - currInterval.range.min - (finalIndex === 0 ? 0n : 1n));
```

Comment:
This `- 1n` adjustment is trying to make up for the earlier off‑by‑one,
but once we count the full intervals correctly we don’t need any patch
at all.  For an inclusive range the slice from `range.min` up to
`amountIn` (inclusive) has `amountIn - range.min + 1` units.

```suggestion
// inclusive slice: count units from min up to amountIn
const size = amountIn - currInterval.range.min + 1n;
amountOut += currInterval.price * size;

Reply:
This is out of date.